### PR TITLE
Add credentials logic

### PR DESCRIPTION
**Pull request by alebedev87**
What's inside:
- Handling of the credentials/config files for all the 5 supported providers. The way the secret is consumed is inspired by [the corresponding upstream guides](https://github.com/kubernetes-sigs/external-dns/tree/master/docs/tutorials). GCP credentials are inspired by Mate (father of ExternalDNS) [docs](https://github.com/linki/mate/blob/v0.6.2/examples/google/README.md#permissions).
- Unit tests for the desired externalDNS deployment was rewritten to cover more use cases. I tried to test any nominal and corner case for each provider and each API option (hostname annotation policy, fqdn template, clusterip publishing, multiple zones, etc.)
- Each service type filter is added as a separate flag now (many comma separated types in one flag doesn't work)

Main files:
- pkg/operator/controller/externaldns/pod.go (handling of credentials)
- pkg/operator/controller/externaldns/deployment_test.go (desired deployment tests)

What's next:
- Secret specified in ExternalDNS CR should be copied into the operand namespace. A dedicated controller is needed for this.

### CODE REVIEW

File: pkg/operator/controller/externaldns/deployment.go [hunk 1]
```diff
@@ -65,7 +69,12 @@ var sourceStringTable = map[operatorv1alpha1.ExternalDNSSourceType]string{
 func (r *reconciler) ensureExternalDNSDeployment(ctx context.Context, namespace, image string, serviceAccount *corev1.ServiceAccount, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
 	nsName := types.NamespacedName{Namespace: namespace, Name: controller.ExternalDNSResourceName(externalDNS)}
 
-	desired, err := desiredExternalDNSDeployment(namespace, image, serviceAccount, externalDNS)
+	secret, err := r.extractCredentialsSecret(ctx, externalDNS)
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to extract credentials secret: %w", err)
+	}
+
+	desired, err := desiredExternalDNSDeployment(namespace, image, serviceAccount, secret, externalDNS)
```
Comment (Miciah) [inline pkg/operator/controller/externaldns/deployment.go:77]: `desiredExternalDNSDeployment` just needs the secret's name so that it can put the name in the volume and env specs, right?  Is there a need to verify that the secret exists and pass it to `desiredExternalDNSDeployment` instead of just passing in the name of the secret?

File: pkg/operator/controller/externaldns/deployment.go [hunk 2]
```diff
@@ -105,8 +114,18 @@ func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, nsName ty
...
 
+// extractCredentialsSecret retieves the contents of the credentials' secret
+func (r *reconciler) extractCredentialsSecret(ctx context.Context, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Secret, error) {
+	secretName := externalDNSCredentialsSecretName(externalDNS)
+	secret := &corev1.Secret{}
+	if err := r.client.Get(ctx, secretName, secret); err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
```
Comment (Miciah) [inline pkg/operator/controller/externaldns/deployment.go:125]: It appears that the only provider-specific logic for credentials is in determining the name of the secret.  Is that correct?

If so, the fact that this logic is almost completely provider-agnostic begs the question of whether the API field needs to be provider-specific (e.g., `Credentials` in `ExternalDNSAWSProviderOptions`) or whether it could be simplified by replacing `Credentials` in `ExternalDNSAWSProviderOptions`, `ExternalDNSGCPProviderOptions`, and `ExternalDNSInfobloxProviderOptions` and `ConfigFile` in `ExternalDNSAzureProviderOptions` and `ExternalDNSBlueCatProviderOptions` with a single `Credentials` field under `ExternalDNSProvider`.  

Is that something worth considering?  Is there some provider where the credentials config is wildly different (in which case presumably this logic will need to be significantly changed)?  (Maybe we already discussed this?)

I suppose one counterargument is that even if each provider needs a credentials/config secret, the keys vary from one provider to the next, so maybe using the same API field for all providers would be misleading.  :thinking:

File: pkg/operator/controller/externaldns/deployment.go [hunk 3]
```diff
@@ -105,8 +114,18 @@ func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, nsName ty
 	return true, depl, nil
 }
 
+// extractCredentialsSecret retieves the contents of the credentials' secret
```
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/operator/controller/externaldns/deployment.go:117]: As the function only has one caller maybe it is simpler and more straightforward to just have this code inline at the call site. It would remove the hard part (i.e., naming!).

File: pkg/operator/controller/externaldns/deployment.go (hunk 2 above)
Comment (frobware) [reply pkg/operator/controller/externaldns/deployment.go:125]: > I suppose one counterargument is that even if each provider needs a credentials/config secret, the keys vary from one provider to the next, so maybe using the same API field for all providers would be misleading.

I wonder if this is the reality. The times I've used cloudflare there's an account key and an API token.

**Review by frobware — CHANGES_REQUESTED** [review]

File: pkg/operator/controller/externaldns/deployment.go (hunk 1 above)
Comment (alebedev87) [reply pkg/operator/controller/externaldns/deployment.go:77]: The secret is supposed to be brought by the other controller: `credentials-secret`. In the worth case scenario (secret provider in CR doesn't exist) we end up with a POD which cannot start because of a missing secret. I think it's more visible (the the creator of CR) than not creating a deployment if the secret is not there.

File: api/v1alpha1/externaldns_types.go [hunk 4]
```diff
@@ -258,14 +259,20 @@ type ExternalDNSGCPProviderOptions struct {
...
 	// See
 	// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys
 	// and https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/gke.md
 	// for more information
 	//
+	// NOTE
+	//
+	// This field is not supported by ExternalDNS.
+	// In order to be able to access GCP DNS, ExternalDNS' service account has to be given with GCP DNS admin IAM role.
+	// There are multiple ways of doing this, see the above mentioned tutorial.
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply api/v1alpha1/externaldns_types.go:274]: We did determine that it's possible, it was not in the doc though. I put this into the PR description:

> GCP credentials are inspired by Mate (father of ExternalDNS) docs.

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 5]
```diff
@@ -30,77 +30,780 @@ import (
 )
 
 const (
-	namespace = "externaldns"
-	name      = "test"
-	image     = "bitname/external-dns:latest"
+	namespace   = "externaldns"
+	name        = "test"
+	image       = "bitname/external-dns:latest"
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply pkg/operator/controller/externaldns/deployment_test.go:35]: It doesn't really matter for these tests as we are not running an instance of ExternalDNS. It could be any name. We just check the logic of the functions which are supposed to ensure the deployment. I think that's why Steve made this `bitname` (not `bitnami`) joke :)

File: pkg/operator/controller/externaldns/pod.go [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+		args = append(args, fmt.Sprintf("--fqdn-template=%s", b.externalDNS.Spec.Source.FQDNTemplate))
+	}
+
+	//TODO: Add logic for the CRD source.
+
+	container.Args = append(container.Args, args...)
+}
+
+// fillProviderSpecificFields fills the fields specific to the provider of given ExternalDNS
+func (b *externalDNSContainerBuilder) fillProviderSpecificFields(seq int, zone string, container *corev1.Container) {
```
[the remark this answers is not in the document]
Comment (alebedev87) [reply pkg/operator/controller/externaldns/pod.go:177]: yes, I just wanted this function to look like it's provider agnostic friend: `fillProviderAgnosticFields()`. Will remove not used fields.

File: api/v1alpha1/externaldns_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alebedev87) [reply api/v1alpha1/externaldns_types.go:274]: Sorry I got it wrong, you are right I put it before I found the way out, will change this.

File: pkg/operator/controller/externaldns/deployment.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alebedev87) [reply pkg/operator/controller/externaldns/deployment.go]: Inlined the function as suggested by @frobware .

File: api/v1alpha1/externaldns_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alebedev87) [reply api/v1alpha1/externaldns_types.go]: Removed this note and the link to the doc which is GKE specific.

File: pkg/operator/controller/externaldns/deployment.go (hunk 2 above)
Comment (alebedev87) [reply pkg/operator/controller/externaldns/deployment.go]: 1. Infoblox provider has some fields additional to the credentials secret: https://github.com/openshift/external-dns-operator/blob/main/api/v1alpha1/externaldns_types.go#L342-L356
2. You are right @Miciah, the keys in the secret are different for different providers.
3. Also, some providers expect full config file (Azure, GCP, BlueCat) and some others just credentials (AWS, Infoblox). Which means that the potential consolidated name has to be well chosen to work well with both.

**Comment by alebedev87** [comment]
Rebased from `main`

**Review by frobware — APPROVED** [review]

**Comment by frobware** [comment]
/lgtm

We can iterate from here with e2e testing. Thanks!
